src/Features/Core/Diagnostics/AnalyzerHelper.cs

```diff
@@ -34,43 +34,41 @@ public static bool IsCompilerAnalyzer(DiagnosticAnalyzer analyzer)
+            Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException = (ex, a, diagnostic) =>
+                OnAnalyzerException_NoTelemetryLogging(ex, a, diagnostic, hostDiagnosticUpdateSource);
 
-        public static Action<Diagnostic> GetAddExceptionDiagnosticDelegate(DiagnosticAnalyzer analyzer, AbstractHostDiagnosticUpdateSource hostDiagnosticUpdateSource, Workspace workspace)
-        {
-            return diagnostic =>
-                hostDiagnosticUpdateSource?.ReportAnalyzerDiagnostic(analyzer, diagnostic, workspace, null);
+            return AnalyzerExecutor.CreateForSupportedDiagnostics(onAnalyzerException, cancellationToken);
         }
 
-        public static AnalyzerExecutor GetAnalyzerExecutorForSupportedDiagnostics(
-            DiagnosticAnalyzer analyzer, 
+        internal static void OnAnalyzerException_NoTelemetryLogging(
+            Exception e,
+            DiagnosticAnalyzer analyzer,
+            Diagnostic diagnostic,
             AbstractHostDiagnosticUpdateSource hostDiagnosticUpdateSource,
-            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException, 
-            CancellationToken cancellationToken)
+            Project projectOpt = null,
+            bool testOnly_RethrowAnalyzerException = false)
         {
-            var addExceptionDiagnostic = GetAddExceptionDiagnosticDelegate(analyzer, hostDiagnosticUpdateSource, hostDiagnosticUpdateSource?.Workspace);
+            if (diagnostic != null)
+            {
+                hostDiagnosticUpdateSource?.ReportAnalyzerDiagnostic(analyzer, diagnostic, hostDiagnosticUpdateSource?.Workspace, projectOpt);
+            }
 
-            // Skip telemetry logging if the exception is thrown as we are computing supported diagnostics and
-            // we can't determine if any descriptors support getting telemetry without having the descriptors.
-            return AnalyzerExecutor.CreateForSupportedDiagnostics(addExceptionDiagnostic, continueOnAnalyzerException, cancellationToken);
-        }
-        
-        public static AnalyzerExecutor GetAnalyzerExecutor(
-            DiagnosticAnalyzer analyzer, 
-            AbstractHostDiagnosticUpdateSource hostDiagnosticUpdateSource, 
-            Project project, 
-            Compilation compilation, 
-            Action<Diagnostic> addDiagnostic,
-            AnalyzerOptions analyzerOptions,
-            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
-            CancellationToken cancellationToken)
-        {
-            var addExceptionDiagnostic = GetAddExceptionDiagnosticDelegate(analyzer, hostDiagnosticUpdateSource, project);
-            return AnalyzerExecutor.Create(compilation, analyzerOptions, addDiagnostic, addExceptionDiagnostic, continueOnAnalyzerException, cancellationToken);
+            if (testOnly_RethrowAnalyzerException)
+            {
+                throw e;
```

Comment:
@jaredpar AnalyzerExecutor is shared code used by IDE driver and compiler driver. IDE wants to log non-fatal exceptions hence the delegate must take the raised exception. Compiler layer ignores all parameters except the diagnostic, and just logs the diagnostic. Given that there is just a single entry point in whole of CodeAnalysis layer that initializes this delegate and creates AnalyzerExecutor (see AnalyzerDriver.Create method), we should be fairly certain that any explicit callbacks into analyzer are always guarded from exceptions.